src/basic/util.c

```diff
@@ -53,29 +53,46 @@ int prot_from_flags(int flags) {
 
 bool in_initrd(void) {
         int r;
+        const char *e;
 
         if (saved_in_initrd >= 0)
                 return saved_in_initrd;
 
-        /* We make two checks here:
+        /* We have two checks here:
          *
          * 1. the flag file /etc/initrd-release must exist
          * 2. the root file system must be a memory file system
          *
          * The second check is extra paranoia, since misdetecting an
          * initrd can have bad consequences due the initrd
          * emptying when transititioning to the main systemd.
+         *
+         * If env var $SYSTEMD_IN_INITRD is not set or set to "auto",
+         * both checks are used. If it's set to "lenient", only check
+         * 1 is used. If set to a booleen value, then the boolean
+         * value is returned.
          */
 
-        r = getenv_bool_secure("SYSTEMD_IN_INITRD");
-        if (r < 0 && r != -ENXIO)
+        e = secure_getenv("SYSTEMD_IN_INITRD");
+        if (e) {
+                r = parse_boolean(e);
+                if (r >= 0) {
+                        saved_in_initrd = r > 0;
+                        return saved_in_initrd;
+                }
+
+                if (streq(e, "auto"))
+                        goto check_all;
+                else if (streq(e, "lenient"))
+                        goto check_etc;
+
                 log_debug_errno(r, "Failed to parse $SYSTEMD_IN_INITRD, ignoring: %m");
+        }
 
-        if (r >= 0)
-                saved_in_initrd = r > 0;
-        else
-                saved_in_initrd = access("/etc/initrd-release", F_OK) >= 0 &&
-                                  path_is_temporary_fs("/") > 0;
+check_all:
+        saved_in_initrd = path_is_temporary_fs("/") > 0;
+check_etc:
+        saved_in_initrd = saved_in_initrd && access("/etc/initrd-release", F_OK) >= 0;
 
         return saved_in_initrd;
```

Comment:
goto, eh?
it's kinda ugly, that in the lenient case you rely on saved_in_initrd = -1 resolving to true...

what not this:

```c
bool lenient = false;

e = secure_getenv(…);
if (e) {
        if (streq(e, "lenient"))
                lenient = true;
        else if (!streq(e, "auto")) {
                r = parse_boolean(e);
                if (r < 0) 
                        log_debug_errno(r, …);
                else {
                        saved_in_initrd = r;
                        return saved_in_initrd;
                }
        }
}

saved_in_initrd = (lenient || path_is_temporary_fs("/") > 0) && access(…) >= 0;
…
```

i.e. one additional bool variable, but no gotos. Which is think preferable.

I mean, gotos are OK in our codebase, but if we can avoid them, still better. And an expressive local variable is prettier. And the compiler should be smart enough to optimize it away anyway...

Reply:
Good suggestion, I hesitated about using a goto or another variable, let me avoid these "goto"s.